# Exploring eBay Car Sales Data

## Description of the Project

eBay Kleinanzeigen is a classifieds section of the German eBay website. 

The dataset was originally scraped and uploaded to Kaggle by user [orgesleka](https://www.kaggle.com/orgesleka).
The original dataset isn't available on Kaggle anymore, but you can find it [here](https://data.world/data-society/used-cars-data). 

We will be using a modified version of the dataset by Dataquest, thus it won't be available in this repository.

## Project Goal

The aim of this project is to clean the data and analyse the used car listings in order to answer the following questions. 

1) Which brands are most frequently listed on the site?
2) What is the average price for each brand?
3) What is the average mileage for each brand?
4) What registration years are most frequently advertised?
5) Which car brands most often have unrepaired damage?

## Processing and Analysing

Importing pandas and NumPy:

In [1]:
import pandas as pd
import numpy as np

Importing the CSV:

In [2]:
autos = pd.read_csv("autos.csv",encoding="Latin-1")
print(autos.info())
print(autos.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

We can see the following information:
- There are 20 columns of different attributes
- There are 50,000 rows/entries/observations.
- The columns names use camelcase instead of the preferred snakecase for Python.
- Some columns have null values
- Columns referring to dates are not in a datetime format
- The odometer columns contains commas and km, and thus is not an int datatype
- The price columns contains commas and dollar signs, and thus is not an int datatype
- Many columns contain null values

### Cleaning the Column Names

Listing the column names:

In [3]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

Creating a map and renaming the columns:

In [4]:
mapping_dict = {
    "dateCrawled":"date_crawled",
    "offerType":"offer_type",
    "vehicleType":"vehicle_type",
    "yearOfRegistration":"registration_year",
    "powerPS":"power_ps",
    "monthOfRegistration":"registration_month",
    "fuelType":"fuel_type",
    "notRepairedDamage":"unrepaired_damage",
    "dateCreated":"ad_created",
    "nrOfPictures":"number_of_pictures",
    "postalCode":"postal_code",
    "lastSeen":"last_seen"}

## renaming the columns using the mapping dictionary to find the original name and replace it with the new name
autos.rename(columns = mapping_dict, inplace=True)
autos.columns

## autos.columns = autos.columns.str.lower() could also be used to just make them all lowercase

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'number_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

Describing the columns:

In [5]:
autos.describe(include='all')
## include='all' will include columns that are categorical

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  number_of_pictures   postal_code  \
count                 50000             50000.0  50000.000000   
unique                   76                 NaN           NaN   
top     2016-04-03 00:00:00                 NaN           NaN   
freq                   1946                 NaN           NaN   
mean                    NaN                 0.0  50813.627300   
std                     NaN                 0.0  25779.747957   
min                     NaN                 0.0   1067.000000   
25%                     NaN                 0.0  30451.000000   
50%                     NaN                 0.0  49577.000000   
75%                     NaN                 0.0  71540.000000   
max                     NaN                 0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

### Converting Columns to Numeric

The price and odometer columns are object datatypes can be converted to int datatypes to be analysed.

Converting the odometer column:

In [6]:
## changing the column name
autos.rename({"odometer":"odometer_km"}, axis=1, inplace=True)

## removing string elements from values
autos["odometer_km"] = (autos["odometer_km"]
                            .str.replace(",","")
                            .str.replace("km","")
                            .astype(int)
                       )

Checking values:

In [7]:
autos.columns
autos["odometer_km"].value_counts()

odometer_km
150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: count, dtype: int64

Converting the price column:

In [8]:
## removing string elements from values
autos["price"] = (autos["price"]
                            .str.replace(",","")
                            .str.replace("$","")
                            .astype(int)
                       )

Checking values:

In [9]:
autos["price"].value_counts()

price
0        1421
500       781
1500      734
2500      643
1000      639
         ... 
414         1
79933       1
5198        1
18890       1
16995       1
Name: count, Length: 2357, dtype: int64

Now that these are datatype int, we should use describe to see the statistics of them:

In [10]:
autos["price"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

Changing the float_format to not show scientific notation:

In [11]:
pd.options.display.float_format = '{:.2f}'.format
autos["price"].describe()

count      50000.00
mean        9840.04
std       481104.38
min            0.00
25%         1100.00
50%         2950.00
75%         7200.00
max     99999999.00
Name: price, dtype: float64

Strangely, there is a max price of 99999999 and a min price of 0.

In [12]:
autos_prices = autos.sort_values("price", ascending=False)
autos_prices.head(20)

date_crawled                                               name  \
39705  2016-03-22 14:58:27                        Tausch_gegen_gleichwertiges   
42221  2016-03-08 20:39:05                                  Leasinguebernahme   
27371  2016-03-09 15:45:47                                         Fiat_Punto   
39377  2016-03-08 23:53:51                        Tausche_volvo_v40_gegen_van   
47598  2016-03-31 18:56:54  Opel_Vectra_B_1_6i_16V_Facelift_Tuning_Showcar...   
2897   2016-03-12 21:50:57   Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000   
24384  2016-03-21 13:57:51                            Schlachte_Golf_3_gt_tdi   
11137  2016-03-29 23:52:57  suche_maserati_3200_gt_Zustand_unwichtig_laufe...   
47634  2016-04-04 21:25:21                                        Ferrari_FXX   
7814   2016-04-04 11:53:31                                        Ferrari_F40   
22947  2016-03-22 12:54:19                         Bmw_530d_zum_ausschlachten   
43049  2016-03-21 19:53:52                                      2_VW_Busse_T3   
514    2016-03-17 09:53:08                   Ford_Focus_Turnier_1.6_16V_Style   
37585  2016-03-29 11:38:54                                Volkswagen_Jetta_GT   
36818  2016-03-27 18:37:37                                        Porsche_991   
14715  2016-03-30 08:37:24                 Rolls_Royce_Phantom_Drophead_Coupe   
34723  2016-03-23 16:37:29  Porsche_Porsche_911/930_Turbo_3.0__deutsche_Au...   
35923  2016-04-03 07:56:23  Porsche_911_Targa_Exclusive_Edition__1_von_15_...   
12682  2016-03-28 22:48:01  Porsche_GT3_RS__PCCB__Lift___grosser_Exklusiv_...   
47337  2016-04-05 10:25:38                                    BMW_Z8_roadster   

       seller offer_type     price   abtest vehicle_type  registration_year  \
39705  privat    Angebot  99999999  control    limousine               1999   
42221  privat    Angebot  27322222  control    limousine               2014   
27371  privat    Angebot  12345678  control          NaN               2017   
39377  privat    Angebot  12345678  control          NaN               2018   
47598  privat    Angebot  12345678  control    limousine               2001   
2897   privat    Angebot  11111111     test    limousine               1973   
24384  privat    Angebot  11111111     test          NaN               1995   
11137  privat    Angebot  10000000  control        coupe               1960   
47634  privat    Angebot   3890000     test        coupe               2006   
7814   privat    Angebot   1300000  control        coupe               1992   
22947  privat    Angebot   1234566  control        kombi               1999   
43049  privat    Angebot    999999     test          bus               1981   
514    privat    Angebot    999999     test        kombi               2009   
37585  privat    Angebot    999990     test    limousine               1985   
36818  privat    Angebot    350000  control        coupe               2016   
14715  privat    Angebot    345000  control       cabrio               2012   
34723  privat    Angebot    299000     test        coupe               1977   
35923  privat    Angebot    295000     test       cabrio               2015   
12682  privat    Angebot    265000  control        coupe               2016   
47337  privat    Angebot    259000     test       cabrio               2001   

         gearbox  power_ps        model  odometer_km  registration_month  \
39705  automatik       224     s_klasse       150000                   9   
42221    manuell       163           c4        40000                   2   
27371        NaN        95        punto       150000                   0   
39377    manuell        95          v40       150000                   6   
47598    manuell       101       vectra       150000                   3   
2897     manuell        48       escort        50000                   3   
24384        NaN         0          NaN       150000                   0   
11137    manuell       368          NaN     

The listings with a price from 999990 and up and above will be removed because they do not appear to represent true values of the models. Listings with a value of 0 will also be removed.

In [13]:
price_bool = autos["price"] < 999990
autos = autos[price_bool]
autos_prices = autos.sort_values("price", ascending=False)
autos_prices.head(10)
autos.shape

(49986, 20)

These 14 rows have been removed.

In [14]:
price_bool = autos["price"] > 0
autos = autos[price_bool]
autos_prices = autos.sort_values("price", ascending=True)
autos_prices.head(10)
autos.shape

(48565, 20)

These 1421 rows have been removed and there are 48565 listings remaining.

### Exploring the Date Columns

The following columns include date/time:
- date_crawled
- registration_year
- registration_month
- ad_created
- last_seen

Exploring the date_crawled column:

In [15]:
autos["date_crawled"].min()

'2016-03-05 14:06:30'

In [16]:
autos["date_crawled"].max()

'2016-04-07 14:36:56'

The data appears to have been scraped over a 1-month period from March 5th until April 7th.

Converting the format to datetime:

In [17]:
autos["date_crawled"] = pd.to_datetime(autos["date_crawled"],format="%Y-%m-%d %H:%M:%S")
autos["date_crawled"]

0       2016-03-26 17:47:46
1       2016-04-04 13:38:56
2       2016-03-26 18:57:24
3       2016-03-12 16:58:10
4       2016-04-01 14:38:50
                ...        
49995   2016-03-27 14:38:19
49996   2016-03-28 10:50:25
49997   2016-04-02 14:44:48
49998   2016-03-08 19:25:42
49999   2016-03-14 00:42:12
Name: date_crawled, Length: 48565, dtype: datetime64[ns]

Creating a column of just the date the ad was scraped:

In [18]:
autos["date_crawled_date"] = autos["date_crawled"].dt.date.astype("datetime64[ns]")
autos["date_crawled_date"]

0       2016-03-26
1       2016-04-04
2       2016-03-26
3       2016-03-12
4       2016-04-01
           ...    
49995   2016-03-27
49996   2016-03-28
49997   2016-04-02
49998   2016-03-08
49999   2016-03-14
Name: date_crawled_date, Length: 48565, dtype: datetime64[ns]

Sorting the crawl date chronologically and getting a count of ads scraped per day:

In [19]:
autos["date_crawled_date"].sort_values(ascending=True).value_counts(sort=False)

date_crawled_date
2016-03-05    1230
2016-03-06     682
2016-03-07    1749
2016-03-08    1617
2016-03-09    1607
2016-03-10    1563
2016-03-11    1582
2016-03-12    1793
2016-03-13     761
2016-03-14    1775
2016-03-15    1665
2016-03-16    1438
2016-03-17    1536
2016-03-18     627
2016-03-19    1689
2016-03-20    1840
2016-03-21    1815
2016-03-22    1602
2016-03-23    1565
2016-03-24    1425
2016-03-25    1535
2016-03-26    1564
2016-03-27    1510
2016-03-28    1693
2016-03-29    1656
2016-03-30    1636
2016-03-31    1546
2016-04-01    1636
2016-04-02    1723
2016-04-03    1875
2016-04-04    1772
2016-04-05     636
2016-04-06     154
2016-04-07      68
Name: count, dtype: int64

There appears to have been consistent scraping until April 5th when fewer ads were scraped.

### Removing Rows with Erroneous Registration Dates

Exploring the registration_year column:

In [20]:
autos["registration_year"].describe()

count   48565.00
mean     2004.76
std        88.64
min      1000.00
25%      1999.00
50%      2004.00
75%      2008.00
max      9999.00
Name: registration_year, dtype: float64

Of the remaining ads, there are some irregularities. There are ads listing cars from the years 1000 and 9999. We need to determine how many ads exist with incorrect registration years.

In [21]:
autos_year_outliers_low = autos[autos["registration_year"] < 1900]
autos_year_outliers_low.shape

(5, 21)

These 5 ads with registration dates before 1900 can be removed:

In [26]:
## replace the autos df with a version of autos that has been filtered by boolean values where TRUE is if the registration_year is above 1900
autos = autos[autos["registration_year"] > 1899]
autos["registration_year"].sort_values(ascending=True).head()

22659    1910
3679     1910
45157    1910
30781    1910
28693    1910
Name: registration_year, dtype: int64

In [28]:
autos.shape

(48560, 21)

For years beyond 2016:

In [30]:
autos_year_outliers_high = autos[autos["registration_year"] > 2016]
autos_year_outliers_high.head()

date_crawled                                               name  \
10 2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
55 2016-03-07 02:47:54                     Mercedes_E320_AMG_zu_Tauschen!   
65 2016-04-04 19:30:39                      Ford_Fiesta_zum_ausschlachten   
68 2016-04-03 17:36:59  Mini_cooper_s_clubman_/vollausstattung_/_Navi/...   
84 2016-03-27 19:52:54                                     Renault_twingo   

    seller offer_type  price   abtest vehicle_type  registration_year  \
10  privat    Angebot    999     test          NaN               2017   
55  privat    Angebot      1     test          NaN               2017   
65  privat    Angebot    250  control          NaN               2017   
68  privat    Angebot  10990     test          NaN               2017   
84  privat    Angebot    900  control          NaN               2018   

      gearbox  power_ps  ... odometer_km  registration_month  fuel_type  \
10    manuell        90  ...      150000                   4     benzin   
55  automatik       224  ...      125000                   7     benzin   
65    manuell        65  ...      125000                   9     benzin   
68    manuell       174  ...      100000                   0        NaN   
84        NaN        60  ...      150000                   0        NaN   

            brand unrepaired_damage           ad_created number_of_pictures  \
10     volkswagen              nein  2016-03-14 00:00:00                  0   
55  mercedes_benz              nein  2016-03-06 00:00:00                  0   
65           ford               NaN  2016-04-04 00:00:00                  0   
68           mini              nein  2016-04-03 00:00:00                  0   
84        renault               NaN  2016-03-27 00:00:00                  0   

    postal_code            last_seen date_crawled_date  
10        86157  2016-04-07 03:16:21        2016-03-15  
55        22111  2016-03-08 05:45:44        2016-03-07  
65        65606  2016-04-05 12:22:12        2016-04-04  
68        83135  2016-04-05 17:26:26        2016-04-03  
84        40589  2016-04-05 18:46:49        2016-03-27  

[5 rows x 21 columns]

In [34]:
autos_year_outliers_high.shape

(1879, 21)

These 1879 rows can be removed as the registration_year should not be beyond April 2016:

In [35]:
## replace the autos df with a version of autos that has been filtered by boolean values where TRUE is if the registration_year is before 2017
autos = autos[autos["registration_year"] < 2017]
autos["registration_year"].sort_values(ascending=False).head()

39584    2016
12564    2016
29501    2016
14303    2016
942      2016
Name: registration_year, dtype: int64

We need to double-check that the 2016 registration_month values for 2016 are not beyond April.

In [37]:
autos_2016 = autos.loc[(autos["registration_year"] == 2016) & (autos["registration_month"] > 4)]
autos_2016.head(5)

date_crawled                                               name  \
135 2016-03-12 11:00:10  Opel_Meriva_B_Panoramadach__Sitz__und_Lenkradh...   
256 2016-04-03 20:50:38                     Passat_1.9TDI_4Motion_Highline   
295 2016-03-28 03:36:22                                     Privat_anbiter   
307 2016-03-15 22:50:48                                       Giessen_ford   
437 2016-03-25 16:39:01                      Mazda__klima_leder__Alufelgen   

     seller offer_type  price   abtest vehicle_type  registration_year  \
135  privat    Angebot   8500  control          NaN               2016   
256  privat    Angebot   4250     test          NaN               2016   
295  privat    Angebot   1000  control          NaN               2016   
307  privat    Angebot   2800     test          NaN               2016   
437  privat    Angebot    550  control          NaN               2016   

     gearbox  power_ps  ... odometer_km  registration_month  fuel_type  \
135  manuell        81  ...       90000                   8     diesel   
256  manuell       131  ...      150000                  11        NaN   
295  manuell         0  ...      150000                   8     benzin   
307  manuell       109  ...      150000                   8        NaN   
437  manuell         0  ...      150000                   7     benzin   

          brand unrepaired_damage           ad_created number_of_pictures  \
135        opel               NaN  2016-03-12 00:00:00                  0   
256  volkswagen               NaN  2016-04-03 00:00:00                  0   
295        opel               NaN  2016-03-28 00:00:00                  0   
307        ford               NaN  2016-03-15 00:00:00                  0   
437       mazda                ja  2016-03-25 00:00:00                  0   

     postal_code            last_seen date_crawled_date  
135        48147  2016-03-22 14:49:31        2016-03-12  
256        31224  2016-04-03 20:50:38        2016-04-03  
295        44147  2016-03-29 23:18:11        2016-03-28  
307        35396  2016-04-07 05:16:07        2016-03-15  
437        70372  2016-04-06 20:48:19        2016-03-25  

[5 rows x 21 columns]

In [38]:
autos_2016.shape

(618, 21)

These 618 rows can also be removed:

In [39]:
autos_bool = (autos["registration_year"] == 2016) & (autos["registration_month"] > 4)
autos_bool.value_counts()

False    46063
True       618
Name: count, dtype: int64

In [40]:
autos = autos[~autos_bool]

Double checking that 46063 rows exist (48565 remainig - 5(<1900) - 1879(>2016) - 618(>April 2016))

In [41]:
autos.shape

(46063, 21)

All ads listed now only include cars registered from 1900 to 2016 April. However, we can take this one step further. The registration date should always be earlier than the ad listing.

Earlier, we saw that many ads had a registration_month of 0. This likely means that no registration month was included in the ad. These will be updated to 1 so that a new datetime column for registration date works.

In [42]:
autos["registration_month"].value_counts()

registration_month
3     4838
0     4026
6     4023
4     3894
5     3785
7     3652
10    3418
12    3199
9     3162
11    3162
1     3105
8     2949
2     2850
Name: count, dtype: int64

In [43]:
month_bool = autos["registration_month"] == 0
autos.loc[month_bool,"registration_month"] = 1
autos["registration_month"].value_counts()

registration_month
1     7131
3     4838
6     4023
4     3894
5     3785
7     3652
10    3418
12    3199
9     3162
11    3162
8     2949
2     2850
Name: count, dtype: int64

Converting registration details to a datetime column:

In [50]:
## Creating a registration_datetime column that combined the registration year and month.
autos["registration_datetime"] = autos["registration_year"].astype(str) + "-" + autos["registration_month"].astype(str)

## Converting this new column to a datetime format
autos["registration_datetime"] = pd.to_datetime(autos["registration_datetime"], format="%Y-%m")

## Viewing the new column
autos["registration_datetime"].head()

0   2004-03-01
1   1997-06-01
2   2009-07-01
3   2007-06-01
4   2003-07-01
Name: registration_datetime, dtype: datetime64[ns]

Converting ad_created details to a datetime column:

In [ ]:
autos["ad_created"] = pd.to_datetime(autos["ad_created"], format="%Y-%m-%d %H:%M:%S")

## Viewing the new column
autos["ad_created"]

Detemining how many ads were registered before or on the month they were listed vs. registered after.

In [72]:
created_bool = autos["registration_datetime"] <= autos["ad_created"]
created_bool.value_counts()

True     45999
False       64
Name: count, dtype: int64

It appears that 64 ads have registration dates after the ad was created.
Viewing examples:

In [73]:
autos_conflicting_dates = autos[autos["registration_datetime"] > autos["ad_created"]]
autos_conflicting_dates

date_crawled                                           name  \
520   2016-03-07 00:52:52           Gut_gepflegter_Golf_GTI_zu_verkaufen   
3055  2016-03-19 16:54:10                              verkaufe_den_Fiat   
3319  2016-03-24 13:52:14                        Vw_t3_lkw_turbo_diesel.   
3414  2016-03-17 22:06:35                              Vw_Golf_5_1.9_TDI   
3988  2016-03-24 22:52:35  Skoda_Octavia_1z3_1.8_TSI_Klima_Navi_Tempomat   
...                   ...                                            ...   
46820 2016-03-14 07:36:20                       Ford_Fiesta_1.3_5_tuerig   
47459 2016-03-14 02:47:45                                     Ford_Focus   
47460 2016-03-27 11:25:20                                  Schoener_lupo   
47935 2016-03-08 10:38:12                        Bmw_e39_in_guten_zutant   
48302 2016-03-08 23:42:31            Golf_3_zum_schlachten_oder_aufbauen   

       seller offer_type  price   abtest vehicle_type  registration_year  \
520    privat    Angebot   6600     test          NaN               2016   
3055   privat    Angebot    170     test          NaN               2016   
3319   privat    Angebot   2100  control          NaN               2016   
3414   privat    Angebot   3700  control          NaN               2016   
3988   privat    Angebot   7500  control          NaN               2016   
...       ...        ...    ...      ...          ...                ...   
46820  privat    Angebot    600  control          NaN               2016   
47459  privat    Angebot    475     test          NaN               2016   
47460  privat    Angebot    799     test          NaN               2016   
47935  privat    Angebot   1850     test          NaN               2016   
48302  privat    Angebot    249  control          NaN               2016   

       gearbox  power_ps  ... registration_month  fuel_type       brand  \
520    manuell       200  ...                  4     benzin  volkswagen   
3055   manuell         0  ...                  4     benzin        fiat   
3319   manuell        70  ...                  4        NaN  volkswagen   
3414   manuell        90  ...                  4     diesel  volkswagen   
3988   manuell       160  ...                  4     benzin       skoda   
...        ...       ...  ...                ...        ...         ...   
46820  manuell        60  ...                  4        NaN        ford   
47459  manuell        90  ...                  4     benzin        ford   
47460  manuell         0  ...                  4     benzin  volkswagen   
47935  manuell         0  ...                  4     benzin         bmw   
48302  manuell        90  ...                  4     benzin  volkswagen   

      unrepaired_damage ad_created number_of_pictures postal_code  \
520                 NaN 2016-03-06                  0       59939   
3055                NaN 2016-03-19                  0       52379   
3319               nein 2016-03-24                  0       48249   
3414                NaN 2016-03-17                  0       55122   
3988               nein 2016-03-24                  0       26605   
...                 ...        ...                ...         ...   
46820               NaN 2016-03-14                  0       15517   
47459                ja 2016-03-13                  0       71144   
47460               NaN 2016-03-27                  0       65520   
47935              nein 2016-03-08                  0       44575   
48302              nein 2016-03-08                  0       82131   

                 last_seen  date_crawled_date registration_datetime  
520    2016-03-12 16:49:44         2016-03-07            2016-04-01  
3055   2016-03-19 16:54:10         2016-03-19            2016-04-01  
3319   2016-04-07 05:17:29         2016-03-24            2016-04-01  
3414   2016-03-19 16:30:51         2016-03-17            2016-04-01  
3988   2016-04-06 02:17:45         2016-03-24            2016-04-01  
...                    ...           

Removing these rows:

In [74]:
autos = autos[autos["registration_datetime"] <= autos["ad_created"]]
autos.shape

(45999, 22)

All ads remaining should now:
- be registered before the ad creation date
- be registered from 1900
- be listed for more than 0 dollars
- be listed for less than 999990 dollars

### Exploring Price by Brand

Creating a list of brands in the dataset:

In [121]:
list_of_brands = autos["brand"].unique()
list_of_brands

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

Creating a dictionary of brands and how frequently they have been listed:

In [182]:
brand_frequencies = {}
for brand in list_of_brands:
    brand_frequencies[brand] = autos[autos["brand"] == brand].shape[0]
    
pd.DataFrame(pd.Series(brand_frequencies),columns=["Frequency"]).sort_values("Frequency",ascending=False).head()

Frequency
volkswagen          9690
bmw                 5088
opel                4922
mercedes_benz       4455
audi                4001

Determining the proportion of ads for each brand:

In [184]:
brand_proportions = {}

for brand in list_of_brands:
    brand_proportions[brand] = autos[autos["brand"] == brand].shape[0] / autos["brand"].shape[0] * 100

pd.DataFrame(pd.Series(brand_proportions),columns=["Proportion"]).sort_values("Proportion",ascending=False).head()

Proportion
volkswagen          21.07
bmw                 11.06
opel                10.70
mercedes_benz        9.68
audi                 8.70

Volkswagon is the brand of car with the most number of ads, and almost twice that of the second most frequent brand, BMW.

Determining the mean value of cars in ads for each brand:

In [186]:
total_value_by_brand = {}
average_value_by_brand = {}
for brand in list_of_brands:
    total_value_by_brand[brand] = autos.loc[autos["brand"] == brand,"price"].sum()
    average_value_by_brand[brand] = autos.loc[autos["brand"] == brand,"price"].mean()
   
pd.DataFrame(pd.Series(average_value_by_brand),columns=["Average Price"]).sort_values("Average Price",ascending=False).head()

Average Price
porsche              45768.56
land_rover           18893.74
sonstige_autos       12341.44
jeep                 11740.50
jaguar               11635.49

Ads for Porsche models have a higher value, on average.

### Exploring Mileage by Brand

In [188]:
total_mileage_by_brand = {}
average_mileage_by_brand = {}

for brand in list_of_brands:
    total_mileage_by_brand[brand] = autos.loc[autos["brand"] == brand,"odometer_km"].sum()
    average_mileage_by_brand[brand] = autos.loc[autos["brand"] == brand,"odometer_km"].mean()
    

pd.DataFrame(pd.Series(total_mileage_by_brand),columns=["Total Mileage"]).sort_values("Total Mileage",ascending=False).head()

Total Mileage
volkswagen        1245455000
bmw                673990000
opel               635840000
mercedes_benz      582015000
audi               516315000

In [190]:
pd.DataFrame(pd.Series(average_mileage_by_brand),columns=["Average Mileage"]).sort_values("Average Mileage",ascending=False).head()

Average Mileage
saab            144415.58
volvo           137955.08
rover           137459.02
bmw             132466.59
chrysler        132049.69

Combining all this information into one dataframe:

In [225]:
brand_info = pd.DataFrame(list_of_brands, columns=["brand"])

brand_info["frequency"] = brand_info["brand"].map(brand_frequencies)
brand_info["proportion"] = brand_info["brand"].map(brand_proportions)
brand_info["total_value"] = brand_info["brand"].map(total_value_by_brand)
brand_info["average_price"] = brand_info["brand"].map(average_value_by_brand)
brand_info["total_mileage"] = brand_info["brand"].map(total_mileage_by_brand)
brand_info["average_mileage"] = brand_info["brand"].map(average_mileage_by_brand)
brand_info["mileage_per_dollar"] = brand_info["total_mileage"] / brand_info["total_value"]

brand_info.sort_values("mileage_per_dollar", ascending=False)

brand  frequency  proportion  total_value  average_price  \
31          daewoo         70        0.15        73430        1049.00   
37           rover         61        0.13        99092        1624.46   
38        daihatsu        116        0.25       191215        1648.41   
7          renault       2168        4.71      5406797        2493.91   
19            saab         77        0.17       247297        3211.65   
11            opel       4922       10.70     14763837        2999.56   
23            fiat       1178        2.56      3337362        2833.07   
0          peugeot       1367        2.97      4250907        3109.66   
5         chrysler        161        0.35       565782        3514.17   
21      mitsubishi        378        0.82      1294747        3425.26   
39          lancia         50        0.11       168811        3376.22   
4             ford       3207        6.97     12169246        3794.59   
35      alfa_romeo        305        0.66      1260984        4134.37   
25          subaru         98        0.21       392329        4003.36   
27         citroen        641        1.39      2451209        3824.04   
36            lada         27        0.06        72584        2688.30   
33         trabant         65        0.14       116406        1790.86   
12           mazda        697        1.52      2896342        4155.44   
30           honda        360        0.78      1490836        4141.21   
20           volvo        423        0.92      2099162        4962.56   
3            smart        648        1.41      2331231        3597.58   
6             seat        835        1.82      3710094        4443.23   
32          suzuki        274        0.60      1128996        4120.42   
17          nissan        700        1.52      3363498        4805.00   
2       volkswagen       9690       21.07     52802904        5449.22   
15          toyota        586        1.27      3045086        5196.39   
29         hyundai        460        1.00      2501091        5437.15   
26             kia        329        0.72      1973519        5998.54   
24           skoda        759        1.65      4844508        6382.75   
1              bmw       5088       11.06     42628095        8378.16   
8    mercedes_benz       4455        9.68     38700547        8686.99   
28       chevrolet        263        0.57      1766581        6717.04   
16           dacia        121        0.26       724360        5986.45   
9             audi       4001        8.70     37564664        9388.82   
18            jeep        105        0.23      1232753       11740.50   
22          jaguar         73        0.16       849391       11635.49   
14            mini        405        0.88      4314415       10652.88   
10  sonstige_autos        455        0.99      5615356       12341.44   
34      land_rover         97        0.21      1832693       18893.74   
13         porsche        284        0.62     12998270       45768.56   

    total_mileage  average_mileage  mileage_per_dollar  
31        8515000        121642.86              115.96  
37        8385000        137459.02               84.62  
38       13495000        116336.21               70.58  
7       277325000        127917.44               51.29  
19       11120000        144415.58               44.97  
11      635840000        129183.26               43.07  
23      137760000        116943.97               41.28  
0       173405000        126850.77               40.79  
5        21260000        132049.69               37.58  
21       47755000        126335.98               36.88  
39        6095000        121900.00               36.11  
4       397830000        124050.51               32.69  
35       39990000        131114.75               31.71  
25       12370000        126224.49               31.53  
27       76490000        119329.17               31.21  
36        2255000         83518.52               31.07  
33        3545000         54538.46               3

Daewoo, Rover and Daihatsu all have much higher mileage relative to their cost.

In [226]:
autos["odometer_km"].mean()

125428.27017978652

Cars advertised have 123,428 kms on their odometer, on average.

### Determining Which Brands Most Often Have Unrepaired Damage